nutrition_labels/cluster_tech_grants.py

```diff
@@ -0,0 +1,108 @@
+import pickle
+from datetime import datetime
+
+import pandas as pd
+import numpy as np
+
+from wellcomeml.ml import TextClustering
+
+from nutrition_labels.grant_data_processing import clean_grants_data
+
+def get_cluster_keys(cluster):
+    """
+    Get information about each cluster:
+    - keywords
+    - size cluster
+    - centroid of cluster
+    """
+
+    cluster_kws = {}
+    cluster_numbers = {}
+    cluster_centroids = {}
+    for cluster_num in list(set(cluster.cluster_ids)):
+        cluster_idx = [i for i, c_num in enumerate(cluster.cluster_ids) if c_num==cluster_num]
+        cluster_kws[cluster_num] = cluster.cluster_kws[cluster_idx[0]]
+        cluster_numbers[cluster_num] = sum(cluster.cluster_ids==cluster_num)
+        cluster_centroids[cluster_num] = np.mean(cluster.reduced_points[cluster_idx], axis = 0)
+        
+    return cluster_kws, cluster_centroids, cluster_numbers
+
+def save_cluster(cluster, output_file):
+    with open(output_file, 'wb') as f:
+        pickle.dump(cluster, file=f)
+
+def load_cluster(cluster_file):
+    with open(cluster_file, 'rb') as f:
+        cluster = pickle.load(f)
+    return cluster
+
+if __name__ == '__main__':
+
+    datestamp = datetime.now().date().strftime('%y%m%d')
+    
+    tech_grants = pd.read_csv("data/processed/ensemble_results.csv")
+    grant_data = pd.read_csv("data/raw/wellcome-grants-awarded-2005-2019.csv")
+
+    output_file = f'models/clustering/grants_clusters_{datestamp}.pkl'
+    tech_output_file = f'models/clustering/tech_grants_clusters_{datestamp}.pkl'
+    grant_data_output_file = f'data/processed/cluster_grant_data_{datestamp}.csv'
+
+    # Process grants data for clustering - clean, merge with tech grants, and
+    # remove duplicate 6 digits, but make sure to keep the ones that are identified as tech grants
+    grant_data = clean_grants_data(grant_data)
+    grants_text = grant_data[['Title', 'Grant Programme:Title', 'Description']].agg(
+                '. '.join, axis=1
+                ).tolist()
+    grant_data['Grant Text'] = grants_text
+    tech_grant_ids = tech_grants['Internal ID'].tolist()
+    grant_data['Tech grant?'] = grant_data['Internal ID'].isin(tech_grant_ids)
+    grant_data.sort_values(by=['Tech grant?'], ascending=False, inplace=True)
+    grant_data.drop_duplicates(subset=['Internal ID 6 digit'], inplace=True)
+
```

Comment:
I dont think you made the column 'Internal ID 6 digit' unless its in your version of wellcome-grants-awarded-2005-2019.csv

Reply:
it should be created when `clean_grants_data` is called

Reply:
yep completely my bad I must have skipped that line by accident when I was going through the code